components/raftstore/src/store/fsm/peer.rs

```diff
@@ -238,6 +257,102 @@ impl<E: KvEngine> PeerFsm<E> {
     }
 }
 
+impl BatchRaftCmdRequestBuilder {
+    fn new(raft_entry_max_size: f64) -> BatchRaftCmdRequestBuilder {
+        BatchRaftCmdRequestBuilder {
+            raft_entry_max_size,
+            request: None,
+            batch_req_size: 0,
+            callbacks: vec![],
+        }
+    }
+
+    fn can_batch(&self, req: &RaftCmdRequest, req_size: u32) -> bool {
+        // No batch request whose size exceed 20% of raft_entry_max_size,
+        // so total size of request in batch_raft_request would not exceed
+        // (40% + 20%) of raft_entry_max_size
+        if req.has_admin_request() || f64::from(req_size) > self.raft_entry_max_size * 0.2 {
+            return false;
+        }
+        for r in req.get_requests() {
+            match r.get_cmd_type() {
+                CmdType::Delete | CmdType::Put | CmdType::DeleteRange => (),
```

Comment:
How can `DeleteRange` be batched?

Reply:
I think it could be..... Because it is applied in order.

Reply:
DeleteRange may delete data in ranges and be conflict with previous pending put/delete.